File: services/google/vpcaccess/connector.go
// Copyright 2021 Google LLC. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package vpcaccess

import (
	"context"
	"fmt"

	"google.golang.org/api/googleapi"
	"github.com/GoogleCloudPlatform/declarative-resource-client-library/dcl"
)

type Connector struct {
	Name          *string             `json:"name"`
	Network       *string             `json:"network"`
	IPCidrRange   *string             `json:"ipCidrRange"`
	MinThroughput *int64              `json:"minThroughput"`
	MaxThroughput *int64              `json:"maxThroughput"`
	Project       *string             `json:"project"`
	Location      *string             `json:"location"`
	State         *ConnectorStateEnum `json:"state"`
	SelfLink      *string             `json:"selfLink"`
}

func (r *Connector) String() string {
	return dcl.SprintResource(r)
}

// The enum ConnectorStateEnum.
type ConnectorStateEnum string

// ConnectorStateEnumRef returns a *ConnectorStateEnum with the value of string s
// If the empty string is provided, nil is returned.
func ConnectorStateEnumRef(s string) *ConnectorStateEnum {
	if s == "" {
		return nil
	}

	v := ConnectorStateEnum(s)
	return &v
}

func (v ConnectorStateEnum) Validate() error {
	for _, s := range []string{"STATE_UNSPECIFIED", "READY", "CREATING", "DELETING", "ERROR", "UPDATING"} {
		if string(v) == s {
			return nil
		}
	}
	return &dcl.EnumInvalidError{
		Enum:  "ConnectorStateEnum",
		Value: string(v),
		Valid: []string{},
	}
}

// Describe returns a simple description of this resource to ensure that automated tools
// can identify it.
func (r *Connector) Describe() dcl.ServiceTypeVersion {
	return dcl.ServiceTypeVersion{
		Service: "vpcaccess",
		Type:    "Connector",
		Version: "vpcaccess",
	}
}

const ConnectorMaxPage = -1

type ConnectorList struct {
	Items []*Connector

	nextToken string

	pageSize int32

	project string

	location string
}

func (l *ConnectorList) HasNext() bool {
	return l.nextToken != ""
}

func (l *ConnectorList) Next(ctx context.Context, c *Client) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	if !l.HasNext() {
		return fmt.Errorf("no next page")
	}
	items, token, err := c.listConnector(ctx, l.project, l.location, l.nextToken, l.pageSize)
	if err != nil {
		return err
	}
	l.Items = items
	l.nextToken = token
	return err
}

func (c *Client) ListConnector(ctx context.Context, project, location string) (*ConnectorList, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	return c.ListConnectorWithMaxResults(ctx, project, location, ConnectorMaxPage)

}

func (c *Client) ListConnectorWithMaxResults(ctx context.Context, project, location string, pageSize int32) (*ConnectorList, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	items, token, err := c.listConnector(ctx, project, location, "", pageSize)
	if err != nil {
		return nil, err
	}
	return &ConnectorList{
		Items:     items,
		nextToken: token,
		pageSize:  pageSize,

		project: project,

		location: location,
	}, nil
}

func (c *Client) GetConnector(ctx context.Context, r *Connector) (*Connector, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	b, err := c.getConnectorRaw(ctx, r)
	if err != nil {
		if dcl.IsNotFound(err) {
			return nil, &googleapi.Error{
				Code:    404,
				Message: err.Error(),
			}
		}
		return nil, err
	}
	result, err := unmarshalConnector(b, c)
	if err != nil {
		return nil, err
	}
	result.Project = r.Project
	result.Location = r.Location
	result.Name = r.Name
	if dcl.IsZeroValue(result.MinThroughput) {
		result.MinThroughput = dcl.Int64(200)
	}
	if dcl.IsZeroValue(result.MaxThroughput) {
		result.MaxThroughput = dcl.Int64(1000)
	}

	c.Config.Logger.Infof("Retrieved raw result state: %v", result)
	c.Config.Logger.Infof("Canonicalizing with specified state: %v", r)
	result, err = canonicalizeConnectorNewState(c, result, r)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Created result state: %v", result)

	return result, nil
}

func (c *Client) DeleteConnector(ctx context.Context, r *Connector) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	if r == nil {
		return fmt.Errorf("Connector resource is nil")
	}
	c.Config.Logger.Info("Deleting Connector...")
	deleteOp := deleteConnectorOperation{}
	return deleteOp.do(ctx, r, c)
}

// DeleteAllConnector deletes all resources that the filter functions returns true on.
func (c *Client) DeleteAllConnector(ctx context.Context, project, location string, filter func(*Connector) bool) error {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	listObj, err := c.ListConnector(ctx, project, location)
	if err != nil {
		return err
	}

	err = c.deleteAllConnector(ctx, filter, listObj.Items)
	if err != nil {
		return err
	}
	for listObj.HasNext() {
		err = listObj.Next(ctx, c)
		if err != nil {
			return nil
		}
		err = c.deleteAllConnector(ctx, filter, listObj.Items)
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *Client) ApplyConnector(ctx context.Context, rawDesired *Connector, opts ...dcl.ApplyOption) (*Connector, error) {
	c.Config.Logger.Info("Beginning ApplyConnector...")
	c.Config.Logger.Infof("User specified desired state: %v", rawDesired)

	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	// 1.1: Validation of user-specified fields in desired state.
	if err := rawDesired.validate(); err != nil {
		return nil, err
	}

	initial, desired, diffs, err := c.connectorDiffsForRawDesired(ctx, rawDesired, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to create a diff: %w", err)
	}

	// TODO(magic-modules-eng): 2.2 Feasibility check (all updates are feasible so far).

	// 2.3: Lifecycle Directive Check
	var create bool
	var recreate bool
	lp := dcl.FetchLifecycleParams(opts)
	if initial == nil {
		if dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
			return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Creation blocked by lifecycle params: %#v.", desired)}
		}
		create = true
	} else if dcl.HasLifecycleParam(lp, dcl.BlockAcquire) {
		return nil, dcl.ApplyInfeasibleError{
			Message: fmt.Sprintf("Resource already exists - apply blocked by lifecycle params: %#v.", initial),
		}
	} else {
		for _, d := range diffs {
			if d.RequiresRecreate {
				if dcl.HasLifecycleParam(lp, dcl.BlockDestruction) || dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
					return nil, dcl.ApplyInfeasibleError{
						Message: fmt.Sprintf("Infeasible update: (%v) would require recreation.", d),
					}
				}
				recreate = true
			}
			if dcl.HasLifecycleParam(lp, dcl.BlockModification) {
				return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Modification blocked, diff (%v) unresolvable.", d)}
			}
		}
	}

	// 2.4 Imperative Request Planning
	var ops []connectorApiOperation
	if create {
		ops = append(ops, &createConnectorOperation{})
	} else if recreate {

		ops = append(ops, &deleteConnectorOperation{})

		ops = append(ops, &createConnectorOperation{})
		// We should re-canonicalize based on a nil existing resource.
		desired, err = canonicalizeConnectorDesiredState(rawDesired, nil)
		if err != nil {
			return nil, err
		}
	} else {
		for _, d := range diffs {
			ops = append(ops, d.UpdateOp)
		}
	}
	c.Config.Logger.Infof("Created plan: %#v", ops)

	// 2.5 Request Actuation
	for _, op := range ops {
		c.Config.Logger.Infof("Performing operation %#v", op)
		if err := op.do(ctx, desired, c); err != nil {
			c.Config.Logger.Infof("Failed operation %#v: %v", op, err)
			return nil, err
		}
		c.Config.Logger.Infof("Finished operation %#v", op)
	}

	// 3.1, 3.2a Retrieval of raw new state & canonicalization with desired state
	c.Config.Logger.Info("Retrieving raw new state...")
	rawNew, err := c.GetConnector(ctx, desired.urlNormalized())
	if err != nil {
		return nil, err
	}

	// Get additional values from the first response.
	// These values should be merged into the newState above.
	if len(ops) > 0 {
		lastOp := ops[len(ops)-1]
		if o, ok := lastOp.(*createConnectorOperation); ok {
			if r, hasR := o.FirstResponse(); hasR {

				c.Config.Logger.Info("Retrieving raw new state from operation...")

				fullResp, err := unmarshalMapConnector(r, c)
				if err != nil {
					return nil, err
				}

				rawNew, err = canonicalizeConnectorNewState(c, rawNew, fullResp)
				if err != nil {
					return nil, err
				}
			}
		}
	}

	c.Config.Logger.Infof("Canonicalizing with raw desired state: %v", rawDesired)
	// 3.2b Canonicalization of raw new state using raw desired state
	newState, err := canonicalizeConnectorNewState(c, rawNew, rawDesired)
	if err != nil {
		return nil, err
	}

	c.Config.Logger.Infof("Created canonical new state: %v", newState)
	// 3.3 Comparison of the new state and raw desired state.
	// TODO(magic-modules-eng): EVENTUALLY_CONSISTENT_UPDATE
	newDesired, err := canonicalizeConnectorDesiredState(rawDesired, newState)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Diffing using canonicalized desired state: %v", newDesired)
	newDiffs, err := diffConnector(c, newDesired, newState)
	if err != nil {
		return nil, err
	}

	if len(newDiffs) == 0 {
		c.Config.Logger.Info("No diffs found. Apply was successful.")
	} else {
		c.Config.Logger.Infof("Found diffs: %v", newDiffs)
		diffMessages := make([]string, len(newDiffs))
		for i, d := range newDiffs {
			diffMessages[i] = fmt.Sprintf("%v", d)
		}
		return newState, dcl.DiffAfterApplyError{Diffs: diffMessages}
	}
	c.Config.Logger.Info("Done Apply.")
	return newState, nil
}

File: python/services/vpcaccess/connector.py
# Copyright 2021 Google LLC. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from connector import channel
from google3.cloud.graphite.mmv2.services.google.vpc_access import connector_pb2
from google3.cloud.graphite.mmv2.services.google.vpc_access import connector_pb2_grpc

from typing import List


class Connector(object):
    def __init__(
        self,
        name: str = None,
        network: str = None,
        ip_cidr_range: str = None,
        min_throughput: int = None,
        max_throughput: int = None,
        project: str = None,
        location: str = None,
        state: str = None,
        self_link: str = None,
        service_account_file: str = "",
    ):

        channel.initialize()
        self.name = name
        self.network = network
        self.ip_cidr_range = ip_cidr_range
        self.min_throughput = min_throughput
        self.max_throughput = max_throughput
        self.project = project
        self.location = location
        self.service_account_file = service_account_file

    def apply(self):
        stub = connector_pb2_grpc.VpcaccessConnectorServiceStub(channel.Channel())
        request = connector_pb2.ApplyVpcaccessConnectorRequest()
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.network):
            request.resource.network = Primitive.to_proto(self.network)

        if Primitive.to_proto(self.ip_cidr_range):
            request.resource.ip_cidr_range = Primitive.to_proto(self.ip_cidr_range)

        if Primitive.to_proto(self.min_throughput):
            request.resource.min_throughput = Primitive.to_proto(self.min_throughput)

        if Primitive.to_proto(self.max_throughput):
            request.resource.max_throughput = Primitive.to_proto(self.max_throughput)

        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        request.service_account_file = self.service_account_file

        response = stub.ApplyVpcaccessConnector(request)
        self.name = Primitive.from_proto(response.name)
        self.network = Primitive.from_proto(response.network)
        self.ip_cidr_range = Primitive.from_proto(response.ip_cidr_range)
        self.min_throughput = Primitive.from_proto(response.min_throughput)
        self.max_throughput = Primitive.from_proto(response.max_throughput)
        self.project = Primitive.from_proto(response.project)
        self.location = Primitive.from_proto(response.location)
        self.state = ConnectorStateEnum.from_proto(response.state)
        self.self_link = Primitive.from_proto(response.self_link)

    def delete(self):
        stub = connector_pb2_grpc.VpcaccessConnectorServiceStub(channel.Channel())
        request = connector_pb2.DeleteVpcaccessConnectorRequest()
        request.service_account_file = self.service_account_file
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.network):
            request.resource.network = Primitive.to_proto(self.network)

        if Primitive.to_proto(self.ip_cidr_range):
            request.resource.ip_cidr_range = Primitive.to_proto(self.ip_cidr_range)

        if Primitive.to_proto(self.min_throughput):
            request.resource.min_throughput = Primitive.to_proto(self.min_throughput)

        if Primitive.to_proto(self.max_throughput):
            request.resource.max_throughput = Primitive.to_proto(self.max_throughput)

        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        response = stub.DeleteVpcaccessConnector(request)

    @classmethod
    def list(self, project, location, service_account_file=""):
        stub = connector_pb2_grpc.VpcaccessConnectorServiceStub(channel.Channel())
        request = connector_pb2.ListVpcaccessConnectorRequest()
        request.service_account_file = service_account_file
        request.Project = project

        request.Location = location

        return stub.ListVpcaccessConnector(request).items

    def to_proto(self):
        resource = connector_pb2.VpcaccessConnector()
        if Primitive.to_proto(self.name):
            resource.name = Primitive.to_proto(self.name)
        if Primitive.to_proto(self.network):
            resource.network = Primitive.to_proto(self.network)
        if Primitive.to_proto(self.ip_cidr_range):
            resource.ip_cidr_range = Primitive.to_proto(self.ip_cidr_range)
        if Primitive.to_proto(self.min_throughput):
            resource.min_throughput = Primitive.to_proto(self.min_throughput)
        if Primitive.to_proto(self.max_throughput):
            resource.max_throughput = Primitive.to_proto(self.max_throughput)
        if Primitive.to_proto(self.project):
            resource.project = Primitive.to_proto(self.project)
        if Primitive.to_proto(self.location):
            resource.location = Primitive.to_proto(self.location)
        return resource


class ConnectorStateEnum(object):
    @classmethod
    def to_proto(self, resource):
        if not resource:
            return resource
        return connector_pb2.VpcaccessConnectorStateEnum.Value(
            "VpcaccessConnectorStateEnum%s" % resource
        )

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return resource
        return connector_pb2.VpcaccessConnectorStateEnum.Name(resource)[
            len("VpcaccessConnectorStateEnum") :
        ]


class Primitive(object):
    @classmethod
    def to_proto(self, s):
        if not s:
            return ""
        return s

    @classmethod
    def from_proto(self, s):
        return s
